src/main/java/io/vertx/uritemplate/impl/UriTemplateImpl.java

```diff
@@ -11,736 +11,108 @@
-    }
-  }
-
-  private static final CharObjectMap<Operator> mapping;
-
-  static {
-    CharObjectMap<Operator> m = new CharObjectHashMap<>();
-    for (Operator op : Operator.values()) {
-      for (char ch : op.so.chars) {
-        m.put(ch, op);
       }
     }
-    mapping = m;
-  }
-
-  public static final class Literals extends Term {
-    private final String value;
-    private Literals(String value) {
-      this.value = value;
-    }
+    return expandToString(variables);
   }
 
-  public static final class Expression extends Term {
-    private final Operator operator;
-    private final List<Varspec> value = new ArrayList<>();
-    public Expression(Operator operator) {
-      this.operator = operator;
-    }
-  }
-
-  public static final class Varspec {
-    public final String varname;
-    public final String decoded;
-    public final int maxLength;
-    public final boolean exploded;
-    private Varspec(String varname, String decoded, int maxLength, boolean exploded) {
-      this.varname = varname;
-      this.decoded = decoded;
-      this.maxLength = maxLength;
-      this.exploded = exploded;
-    }
-  }
-
-  public static class Parser {
-
-    private UriTemplateImpl template;
-    private Expression expression;
-
-    public UriTemplateImpl parseURITemplate(String s) {
-      template = new UriTemplateImpl();
-      if (parseURITemplate(s, 0) != s.length()) {
-        throw new IllegalArgumentException();
-      }
-      for (Term term : template.terms) {
-        if (term instanceof Expression && ((Expression) term).operator == Operator.FUTURE) {
-          throw new IllegalArgumentException("Invalid reserved operator");
-        }
-      }
-      return template;
-    }
-
-    public int parseURITemplate(String s, int pos) {
-      while (true) {
-        int idx = parseLiterals(s, pos);
-        if (idx > pos) {
-          template.terms.add(new Literals(literals.toString()));
-          pos = idx;
-        } else {
-          idx = parseExpression(s, pos);
-          if (idx > pos) {
-            pos = idx;
-          } else {
-            break;
-          }
-        }
-      }
-      return pos;
-    }
-
-    public int parseExpression(String s, int pos) {
-      if (pos < s.length() && s.charAt(pos) == '{') {
-        int idx = pos + 1;
-        Operator operator;
-        if (idx < s.length() && isOperator(s.charAt(idx))) {
-          operator = mapping.get(s.charAt(idx));
-          idx++;
-        } else {
-          operator = Operator.SIMPLE_STRING_EXPANSION;
-        }
-        expression = new Expression(operator);
-        idx = parseVariableList(s, idx);
-        if (idx < s.length() && s.charAt(idx) == '}') {
-          pos = idx + 1;
-        }
-        if (template != null) {
-          template.terms.add(expression);
-        }
-        expression = null;
-      }
-      return pos;
-    }
+  private List<String> extractTokens() {
```

Comment:
The diff in GH looks pretty ugly, anyhow, to respect the current option to throw an exception on missing keys (which is a usability "extra" on top of the original spec), I implemented a single pass extraction of the tokens as an additional validation step.